File: tests/generics/generic.c
#include <stdio.h>
#include <stdlib.h>

struct Entity {
    char *id;
    char* (*get_id)(struct Entity *self);
};

struct LivingBeing {
    struct Entity base;
    void (*breathe)(struct LivingBeing *self);
};

struct Animal {
    struct LivingBeing base;
    char *name;
};

struct Cat {
    struct Animal base;
    void (*speak)(struct Cat *self);
    
    // Nested struct definition
    struct Breed {
        char *species_type;
    } breed;
};

// Nesting profondo di struct
struct Outer {
    struct Inner {
        struct DeepInner {
            int value;
        } deep;
    } inner;
};

void factory() {
    // Local struct
    struct LocalProduct {
        int id;
    };
    struct LocalProduct p = {1};
}

int main() {
    return 0;
}

File: tests/generics/generic.rs
use tree_sitter::Parser

enum NodeType {
    Entity,
    LivingBeing,
    Animal,
    Mammal,
    Cat,
    Chargeable,
    Robot,
}

mod module {
    pub struct ModuleEntity {
        pub id: String,
    }
    
    impl super::Entity for ModuleEntity {
        fn get_id(&self) -> String { self.id.clone() }
    }
}

pub trait Entity {
    fn get_id(&self) -> String;
}

pub trait LivingBeing: Entity {
    fn breathe(&self);
}

fn useless_function() {
    struct LocalEntity {
        pub id: String,
    }
    
    impl Entity for LocalEntity {
        fn get_id(&self) -> String { self.id.clone() }
    }
    
    let entity = LocalEntity { id: "123".to_string() };
    entity.get_id();
}

pub struct Animal {
    pub name: String,
}

impl Entity for Animal {
    fn get_id(&self) -> String { self.name.clone() }
}

impl LivingBeing for Animal {
    fn breathe(&self) { println!("Breathing..."); }
}

pub struct Mammal {
    pub base: Animal,
}

pub struct Cat {
    pub base: Mammal,
}

impl Cat {
    pub fn speak(&self, to_whom: Animal) -> String { println!("Meow"); "meow".to_string() }
    
    pub struct Breed {
        pub species_type: String,
    }
}

pub trait Chargeable: Entity {
    fn charge(&self);
}

pub struct Robot {
    pub id: String,
}

impl Entity for Robot {
    fn get_id(&self) -> String { self.id.clone() }
}

impl Chargeable for Robot {
    fn charge(&self) { println!("Charging..."); }
}

// Moduli come Nesting
pub mod outer {
    pub mod inner {
        pub struct DeepInner;
        
        impl DeepInner {
            pub fn hello(&self) {
                // Local struct in function
                struct Local;
            }
        }
    }
}
